fairscale/experimental/nn/multiprocess_pipe.py

```diff
@@ -85,122 +42,629 @@ def dloss(input_rref: rpc.RRef, target_rref: rpc.RRef) -> rpc.RRef:
+        The user may also specify the input to the first module in this sequence with argument 'first_input'.
+        In this case the module 'first_input' must have been added to the graph previously.
+        """
+        old_modules_len = len(self.modules_list)
+        new_modules_len = len(modules)
+        self.modules_list.extend(modules)
+        # update inputs array
+        self.inputs.append([(self.modules_list.index(first_input), 0)] if first_input is not None else None)
+        for i in range(old_modules_len + 1, old_modules_len + new_modules_len):
+            self.inputs.append([(i - 1, 0)])
+
+    def feed_model_input(self, module: PipelineModule, ind: int = 0) -> None:
+        """Declares the input to a module as the input to the model. In case the model has multiple
+        inputs, the argument 'ind' indicates the index of the model input that is fed to the module.
+        """
+        self.inputs[self._find_or_add(module)] = [(-1, ind)]
+
+    def add_multi_input_layer(self, module: PipelineModule, inputs: List[PipelineModule]) -> None:
+        """Adds a module with multiple inputs to the graph. The modules that provide inpurs to this module
+        must have been added previously to the graph and are listed with argument inputs.
+        """
+        self.inputs[self._find_or_add(module)] = [(self.modules_list.index(m), 0) for m in inputs]
+
+    def fan_out(self, module: PipelineModule, outputs: List[PipelineModule]) -> None:
+        """Feeds outputs of a previously added module to modules specified by argument 'outputs' (so
+        'module' should have at least 'len(outputs)' outputs.
+        Modules in the list 'outputs' are added to the graph if they have not been added previously.
+        """
+        mi = self.modules_list.index(module)
+        for i, m in enumerate(outputs):
+            self.inputs[self._find_or_add(m)] = [(mi, i)]
+
+    def compile_output_users(self) -> None:
+        """Precomputs self.model_input_users and self.output_users for internal use by the pipleine
+        class. These two lists show consumers of inputs to the model, and outputs of each module of
+        the graph. Each consumer is a pair (i, j) which stands for the j'th input to the i'th module
+        in the graph.
+        """
+
+        # TODO: We need to make sure following conditions hold before preparing the graph for the pipeline:
+        #   * the graph has a least one module, and is connected.
+        #   * num_inputs and num_outputs for modules matche list of connections defined in the graph.
+        #   * all inputs to a module should come from model input, or modules with smaller index in
+        #     the graph. This condition is used in implementaion of DistributedPipeline.forward. Even
+        #     if we relax this condition, still need to make sure the graph is acyclic.
+
+        m = len(self.modules_list)
+        self.output_users: List[List[Tuple[int, int, int]]] = [[] for _ in range(m)]
```

Comment:
I would suggest that we create a Python structure to represent this. It's going to be hard to figure out what the three indices stand for down the line.